testsuite/classlibrary/TestServer_boot.sc

```diff
@@ -0,0 +1,132 @@
+TestServer_boot : UnitTest {
+
+	test_waitForBoot {
+		var options = ServerOptions.new;
+		var s = Server(\testserv1, NetAddr("localhost", 57111), options);
+		var vals = List[];
+		var of = OSCFunc({ |msg| vals.add(msg[3]) }, \tr, s.addr);
+		var cond = Condition();
+
+		// waitForBoot tests doWhenBooted implicitly
+		s.waitForBoot({
+
+			var synth = { SendTrig.ar(Impulse.ar(10), 4711, 1) }.play(s);
+			1.wait;
+			synth.free;
+			cond.unhang;
+		});
+
+		cond.hang;
+		this.assert((vals.size > 8 and: vals.every(_ == 1)),
+			"waitForBoot should allow starting a synth."
+		);
+
+		s.quit;
+		s.remove;
+		of.free;
+	}
+
+	test_bootSync {
+		var options = ServerOptions.new;
+		var s = Server(\testserv1, NetAddr("localhost", 57111), options);
+		var vals = List[];
+		var of = OSCFunc({ |msg| vals.add(msg[3]) }, \tr, s.addr);
+		var synth;
+
+		s.bootSync;
+
+		synth = { SendTrig.ar(Impulse.ar(10), 4711, 1) }.play(s);
+		1.wait;
+		synth.free;
+
+		this.assert((vals.size > 8 and: vals.every(_ == 1)),
+			"bootSync should allow starting a synth right after continuing."
+		);
+
+		s.quit;
+		s.remove;
+		of.free;
+	}
+
+	test_allocWhileBooting {
+		var s = Server(\test), done = false, count = 0;
+		var prevNodeID = -1, nodeID = 0, failed = false;
+
+		s.boot;
+		while {
+			done.not and: { failed.not }
+		} {
+			if(s.serverRunning) {
```

Comment:
i think you should increment the counter outside the `if`; otherwise this loop will continue infinitely if `serverRunning` is never set to true for whatever reason.

Reply:
then it does not test asking the server for 1000 nodeIDs, while it is booting, to catch duplicate calls to newAllocators. but you are right, there should be a timeout in case the server never reaches serverRunning. will do that, ok?

Reply:
oh, hm. sounds like you should be waiting for s.serverRunning before this loop.

Reply:
ok, just added a timeout.

Reply:
I appreciate the speed, but it's very confusing now.

Reply:
for instance, you've written that you're giving the server 5 seconds to reach s.serverRunning, but it's actually just 4 because 4000 wait passes + 1000 node allocations, timeoutCount will be below 0.

I'm not totally convinced the timeout is necessary either, because looking at other tests I'm pretty sure something would fail before that. That was a premature response on my part. But the wait and then allocation loop definitely need to be split up for the sake of clarity and correctness.

Reply:
I think you could also get rid of `done` by just checking `count > 1000`

Reply:
ok, none of the other tests checks whether booting really works, 
because the simpler boot tests would then have failed earlier.
but users may also run tests individually, so a timeout is safer.

Is this version clearer?
```
// test with: 
fork { TestServer_boot().test_allocWhileBooting }; 
ServerBoot.add({ "KILLEMALL".postln; Server.killAll; });

	test_allocWhileBooting {
		var s = Server(\test), done = false, count = 0;
		var prevNodeID = -1, nodeID = 0, failed = false;
		var timer = fork {
			5.wait;
			failed = true;
			"% - server boot timed out.".format(thisMethod).warn;
		};

		s.boot;

		while {
			count < 1000 and: { failed.not }
		} {
			if(s.serverRunning) {
				".".post;
				count = count + 1;
				nodeID = s.nextNodeID;
				if(nodeID <= prevNodeID) {
					failed = true;
					"failed: prevNodeID % and nodeID = % "
					.format(prevNodeID, nodeID).warn;
				};
				prevNodeID = nodeID;
			};
			0.001.wait;
		};
		"end of while loop, timer stops...".postln;

		timer.stop;

		this.assert(failed.not,
			"allocating nodeIDs while booting should not produce duplicate nodeIDs."
		);
	}
```